### DISCUSSION TITLE
Specify filename for restclient-reactive multipart messages

### DISCUSSION THREAD

**Discussion by Sopka** [quarkusio/quarkus#31155] [Q&A]
It is already documented how to send multipart messages in the official quarkus documentation at [Sending Multipart messages](https://quarkus.io/guides/rest-client-reactive#sending-multipart-messages).

With the old non-reactive version of the restclient implementation the filename could be set with:

```
import org.jboss.resteasy.plugins.providers.multipart.MultipartFormDataOutput;
[...]
  MultipartFormDataOutput createAttachmentMultipartFormDataOutput(byte[] content, String fileName) {
    MultipartFormDataOutput outputData = new MultipartFormDataOutput();
    outputData.addFormData("file", content, MediaType.APPLICATION_OCTET_STREAM_TYPE, fileName);
    return outputData;
  }
```

The new reactive version is easier and only a PoJo is needed with a few annotations:

```
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class MultipartParameters {
  @RestForm("file")
  @PartFilename("document.txt")
  @PartType(MediaType.APPLICATION_OCTET_STREAM)
  byte[] file;
}
```

Corresponding RestClient:

```
@ApplicationScoped
@RegisterRestClient
public interface RemoteRestClient {
  @Path("/api/attachments")
  String addAttachment(MultipartParameters multipartParameters);
}
```

****Question: But how** can you set the filename dynamically/programatically?**

One possible solution/workaround has been mentioned in this issue: [#26198](https://github.com/quarkusio/quarkus/discussions/26198#discussion-4152367)

An implemented service is cumbersome to implement yourself without the `@RestClient` magic and looks then like:

```
import java.text.MessageFormat;
import java.time.Duration;
import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;
import javax.ws.rs.core.HttpHeaders;
import javax.ws.rs.core.MediaType;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import io.vertx.mutiny.core.Vertx;
import io.vertx.mutiny.core.buffer.Buffer;
import io.vertx.mutiny.ext.web.client.WebClient;
import io.vertx.mutiny.ext.web.multipart.MultipartForm;

@ApplicationScoped
public class AttachmentService {

  private final WebClient client;

  private final String baseUrl;

  @Inject
  public AttachmentService(Vertx vertx,
      @ConfigProperty(name = "quarkus.rest-client.api-id.url") String baseUrl) {
    this.baseUrl = baseUrl;
    this.client = WebClient.create(vertx);
  }

  public void addAttachment(String resourceId, String fileName, byte[] data) {
    MultipartForm multipartForm = MultipartForm.create().binaryFileUpload("file",
        fileName, Buffer.buffer(data), MediaType.APPLICATION_OCTET_STREAM);

    final String url = MessageFormat.format("{0}/api/{1}/attachments",
        this.baseUrl, resourceId);
    this.client.postAbs(url)
        .putHeader(HttpHeaders.CONTENT_TYPE, MediaType.MULTIPART_FORM_DATA)
        .sendMultipartForm(multipartForm)
        .onItem().transform(response -> response.bodyAsString())
        // TODO: .onfailure...
        .subscribe().with(System.out::println);
  }
}
```

**Comment by gefloh** [comment]
Fully agree, that would be helpful. Since I want to keep using the regular RESTEasy Reactive rest client I am currently helping myself by using a `File` object like

```
@ApplicationScoped
@RegisterRestClient
public interface MyRestClient {
    @POST
    @Path("/...")
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    Response upload(@RestForm File file);
}
```

and making sure that the file has the desired filename. For us this seems to work as expected.

**Comment by martaisty** [comment]
@Sopka I'm experiencing the same issue, have you found any solution to that? 

@sberyozkin please take a look. Might it be worth enhancing multipart file sending using rest-client-reactive? When you have `java.io.File`, there is no problem. But when you want to sent pure bytes as a file it seems to be impossible to dynamically specify the filename. Use case: you have file stored in a database and want to send it directly without writing to filesystem

**Comment by sberyozkin** [comment]
CC @geoand @FroMage

**Comment by ikos23** [comment]
We faced it as well. I think it is quite common to read data from DB as byte[] (file data I mean) and it would be nice to be able to specify a name when we do upload. This `@PartFilename("document.txt")` seems to be useless, it's a very rare case when it might be helpful.

**Comment by geoand** [comment]
Just to make sure we are all on the same page, we are talking about specifying the filename when sending a multipart request using the Reactive REST Client, correct?

**Reply by martaisty** [reply]
Yes, exactly👍

**Reply by geoand** [reply]
In that case, we indeed don't seem to have a way to support that... 

Can you please open an issue?

Thanks

**Reply by martaisty** [reply]
Done :+1:  #36130
